Source/NonlinearSolvers/CurlCurlMLMGPC.H

```diff
@@ -0,0 +1,402 @@
+/* Copyright 2024 Debojyoti Ghosh
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef CURL_CURL_MLMG_PC_H_
+#define CURL_CURL_MLMG_PC_H_
+
+#include <AMReX.H>
+#include <AMReX_ParmParse.H>
+#include <AMReX_Array.H>
+#include <AMReX_Vector.H>
+#include <AMReX_MultiFab.H>
+#include <AMReX_MLLinOp.H>
+#include <AMReX_MLCurlCurl.H>
+#include <AMReX_MLMG.H>
+#include <AMReX_GMRES_MLMG.H>
+
+#include "FieldSolver/Fields.H"
+#include "Utils/WarpXConst.H"
+#include "Preconditioner.H"
+
+/**
+ * \brief Curl-curl Preconditioner
+ *
+ *  Preconditioner that solves the curl-curl equation for the E-field, given
+ *  a RHS. Uses AMReX's curl-curl linear operator and multigrid solver.
+ *
+ *  This class is templated on a solution-type class T and an operator class Ops.
+ *
+ *  The Ops class must have the following function:
+ *      + Return number of AMR levels
+ *      + Return the amrex::Geometry object given an AMR level
+ *      + Return hi and lo linear operator boundaries
+ *      + Return the time step factor (theta) for the time integration scheme
+ *
+ *  The T class must have the following functions:
+ *      + Return underlying vector of amrex::MultiFab arrays
+ */
+
+template <class T, class Ops>
+class CurlCurlMLMGPC : public Preconditioner<T,Ops>
+{
+    public:
+
+        using RT = typename T::value_type;
+
+        /**
+         * \brief Default constructor
+         */
+        CurlCurlMLMGPC () = default;
+
+        /**
+         * \brief Default destructor
+         */
+        ~CurlCurlMLMGPC () override = default;
+
+        /**
+         * \brief Define the preconditioner
+         */
+        virtual void Define (const T&, Ops* const) override;
+
+        /**
+         * \brief Update the preconditioner
+         */
+        virtual void Update (const T&) override;
+
+        /**
+         * \brief Apply (solve) the preconditioner given a RHS
+         */
+        virtual void Apply (T&, const T&) override;
+
+        /**
+         * \brief Print parameters
+         */
+        virtual void printParameters() const override;
+
+        /**
+         * \brief Check if the nonlinear solver has been defined.
+         */
+        [[nodiscard]] inline bool IsDefined () const override { return m_is_defined; }
+
+    protected:
+
+        using MFArr = amrex::Array<amrex::MultiFab,3>;
+
+        bool m_is_defined = false;
+
+        bool m_verbose = true;
+        bool m_bottom_verbose = false;
+        bool m_agglomeration = true;
+        bool m_consolidation = true;
+        bool m_use_gmres = false;
+        bool m_use_gmres_pc = true;
+
+        int m_max_iter = 300;
+        int m_max_coarsening_level = 30;
+
+        RT m_atol = 1.0e-16;
+        RT m_rtol = 1.0e-10;
```

Comment:
@JustinRayAngus can I set this to `1e-3`?

Reply:
How about 1.0e-4?

Reply:
Set to `1e-4`.